### PR TITLE
Extract and process test method names 

### PR DESCRIPTION
**Pull request by bryan-lou**
This PR adds the functionality to extract the individual test cases & suites and add them to a column.

### CODE REVIEW

File: pkg/updater/resultstore/resultstore.go [hunk 1]
```diff
@@ -409,20 +422,67 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
...
+				// TODO(sultan-duisenbay): sanitize build target and apply naming config
+				var cell updater.Cell
+				cell.CellID = invocation.InvocationProto.GetId().GetInvocationId()
+				cell.ID = targetID
 				cell.Result = status
-				if cr := customTargetStatus(tg.GetCustomEvaluatorRuleSet(), sar); cr != nil {
-					cell.Result = *cr
+
+				testResults := getTestResults(sar.ActionProto.GetTestAction().GetTestSuite())
+				for _, testResult := range testResults {
```
Comment (michelle192837) [inline pkg/updater/resultstore/resultstore.go:438]: nit: I think this is worth putting in a separate helper method, so it's clear what processing a group does.

File: pkg/updater/resultstore/resultstore.go [hunk 2]
```diff
@@ -379,6 +390,8 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
 	hintTime := time.Unix(0, 0)
 	headers := make([][]string, len(tg.GetColumnHeader()))
 
+	receivers := make(chan updater.InflatedColumn)
```
Comment (michelle192837) [inline pkg/updater/resultstore/resultstore.go:393]: nit: Up to you, but I'd leave channels/parallelization for later. We can do correct behavior first and scale the processing up later once we know all the details work. (that should help the timeout in unit tests too).

File: pkg/updater/resultstore/resultstore_test.go [hunk 3]
```diff
@@ -1132,6 +1132,105 @@ func TestProcessGroup(t *testing.T) {
...
 }
 
+func TestGetTestResults(t *testing.T) {
+	cases := []struct {
+		name      string
+		testsuite *resultstore.TestSuite
+		want      []*resultstore.TestCase
+	}{
+		{
+			name: "all tests",
```
Comment (michelle192837) [inline pkg/updater/resultstore/resultstore_test.go:1142]: This is a good case, though I think a few more to cover edge and complicated cases would be good.

File: pkg/updater/resultstore/resultstore.go [hunk 4]
```diff
@@ -363,10 +363,21 @@ func (t *testResult) properties() []*resultstorepb.Property {
 
 // processGroup will convert grouped invocations into columns
 func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.InflatedColumn {
-	if group == nil || group.Invocations == nil {
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:366]: i would leave the concurrency for later - let's make that code is working correctly in synchronous manner first. So for now, no need for `contexts` and `channels`

File: pkg/updater/resultstore/resultstore.go (hunk 2 above)
Comment (sultan-duisenbay) [reply pkg/updater/resultstore/resultstore.go:393]: +1

File: pkg/updater/resultstore/resultstore.go [hunk 5]
```diff
@@ -379,6 +390,8 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
 	hintTime := time.Unix(0, 0)
 	headers := make([][]string, len(tg.GetColumnHeader()))
 
+	receivers := make(chan updater.InflatedColumn)
+	var sent int
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:394]: ditto: no need for this too for now

File: pkg/updater/resultstore/resultstore.go (hunk 1 above)
Comment (sultan-duisenbay) [reply pkg/updater/resultstore/resultstore.go:438]: +1

File: pkg/updater/resultstore/resultstore.go [hunk 6]
```diff
@@ -409,20 +422,67 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
...
+				cell.ID = targetID
 				cell.Result = status
-				if cr := customTargetStatus(tg.GetCustomEvaluatorRuleSet(), sar); cr != nil {
-					cell.Result = *cr
+
+				testResults := getTestResults(sar.ActionProto.GetTestAction().GetTestSuite())
+				for _, testResult := range testResults {
+					// TODO(bryanlou) targetID + "@TESTGRID@"
+					var methodName string
+					if testResult.GetTestCase() == nil {
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:441]: if testResult.GetTestCase() != nil

File: pkg/updater/resultstore/resultstore.go [hunk 7]
```diff
@@ -409,20 +422,67 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
...
+							cell.Result = statuspb.TestStatus_FAIL
+						default:
+							cell.Result = statuspb.TestStatus_PASS
+						}
+					}
+					if cell.Result == statuspb.TestStatus_PASS_WITH_SKIPS && tg.IgnoreSkip {
+						continue
+					}
+					if len(testResults) <= testMethodLimit {
+						tcName := testResult.GetTestCase().GetCaseName()
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:480]: we should use `methodName` here

File: pkg/updater/resultstore/resultstore.go [hunk 8]
```diff
@@ -440,6 +500,12 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
 				col.Cells[outName] = outCell
 			}
 		}
+		select {
+		case <-ctx.Done():
+			return col
+		case receivers <- *col:
+			sent++
+		}
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:508]: no need for this for now

File: pkg/updater/resultstore/resultstore.go [hunk 9]
```diff
@@ -409,20 +422,67 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
...
 				if !ok {
 					status = statuspb.TestStatus_UNKNOWN
 				}
+				// TODO(sultan-duisenbay): sanitize build target and apply naming config
+				var cell updater.Cell
+				cell.CellID = invocation.InvocationProto.GetId().GetInvocationId()
+				cell.ID = targetID
 				cell.Result = status
-				if cr := customTargetStatus(tg.GetCustomEvaluatorRuleSet(), sar); cr != nil {
-					cell.Result = *cr
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:425]: let's not remove this for now... we should leave it for targets, but maybe add this for test results in a separate PR

File: pkg/updater/resultstore/resultstore.go [hunk 10]
```diff
@@ -409,20 +422,67 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
...
+						case res == resultstorepb.TestCase_CANCELLED:
+							cell.Result = statuspb.TestStatus_PASS_WITH_SKIPS
+						case res == resultstorepb.TestCase_CANCELLED:
+							cell.Result = statuspb.TestStatus_CANCEL
+						case len(testResult.GetTestCase().GetFailures()) > 0 || len(testResult.GetTestCase().GetErrors()) > 0:
+							cell.Result = statuspb.TestStatus_FAIL
+						default:
+							cell.Result = statuspb.TestStatus_PASS
+						}
+					}
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:475]: let's add this in a separate PR

File: pkg/updater/resultstore/resultstore.go [hunk 11]
```diff
@@ -453,6 +519,31 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
 	return col
 }
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:521]: nit: add a comment like:
`getTestResults does xxxx`

File: pkg/updater/resultstore/resultstore_test.go (hunk 3 above)
Comment (sultan-duisenbay) [reply pkg/updater/resultstore/resultstore_test.go:1142]: we should also test other cases: e.g. for example nested testSuites or a testSuite without testCases

**Review by sultan-duisenbay — CHANGES_REQUESTED** [review]

File: pkg/updater/resultstore/resultstore_test.go (hunk 3 above)
Comment (bryan-lou) [reply pkg/updater/resultstore/resultstore_test.go]: done!

File: pkg/updater/resultstore/resultstore.go (hunk 4 above)
Comment (bryan-lou) [reply pkg/updater/resultstore/resultstore.go]: done.

File: pkg/updater/resultstore/resultstore.go (hunk 10 above)
Comment (bryan-lou) [reply pkg/updater/resultstore/resultstore.go]: okay, sounds good

File: pkg/updater/resultstore/resultstore.go [hunk 12]
```diff
@@ -453,6 +457,75 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
...
+		}
+	}
+	return tests
+}
+
+func testMethodLimit(tg *configpb.TestGroup) int {
+	var testMethodLimit int
+	const defaultTestMethodLimit = 20
+	if tg == nil {
+		return defaultTestMethodLimit
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:490]: `return testMethodLimit`

File: pkg/updater/resultstore/resultstore.go [hunk 13]
```diff
@@ -453,6 +457,75 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
...
+		}
+
+		methodName = targetID + "@TESTGRID@" + methodName
+
+		if cell.Result == statuspb.TestStatus_PASS_WITH_SKIPS && tg.IgnoreSkip {
+			continue
+		}
+		if len(testResults) <= testMethodLimit {
+			groupedCells[methodName] = append(groupedCells[methodName], cell)
+		}
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:525]: you should create a separate testResultCell and append it to the groupedCells. you can create one like this:
```
trCell := Cell{
  ID: targetID, // same targetID as the parent TargetResult
  CellID: cell.CellID, // same cellID
  Result: testResultStatus(testResult) // you have to perform mapping of statuses
}
```

File: pkg/updater/resultstore/resultstore.go (hunk 12 above)
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:490]: return testMethodLimit or 0

File: pkg/updater/resultstore/resultstore_test.go [hunk 14]
```diff
@@ -1132,6 +1133,176 @@ func TestProcessGroup(t *testing.T) {
...
+	cases := []struct {
+		name      string
+		testsuite *resultstore.TestSuite
+		want      []*resultstore.Test
+	}{
+		{
+			name: "empty test suite",
+			testsuite: &resultstore.TestSuite{
+				SuiteName: "Empty Test",
+				Tests:     []*resultstore.Test{},
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore_test.go:1146]: if you remove the check on line 463 in `resultstore.go` as I suggested, you don't have to specify an empty slice of Tests here

File: pkg/updater/resultstore/resultstore_test.go [hunk 15]
```diff
@@ -1132,6 +1133,176 @@ func TestProcessGroup(t *testing.T) {
...
+			},
+			want: []*resultstore.Test{
+				{
+					TestType: &resultstore.Test_TestSuite{
+						TestSuite: &resultstore.TestSuite{
+							SuiteName: "Empty Test",
+							Tests:     []*resultstore.Test{},
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore_test.go:1153]: ditto

File: pkg/updater/resultstore/resultstore_test.go [hunk 16]
```diff
@@ -1132,6 +1133,176 @@ func TestProcessGroup(t *testing.T) {
...
+				},
+			},
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			got := getTestResults(tc.testsuite)
+			if diff := reflect.DeepEqual(tc.want, got); diff != true {
+				t.Errorf("getTestResults() differed (-want, +got):\n [%s]\n [%s]", got, tc.want)
+			}
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore_test.go:1301]: use `cmp` library to perform comparison as in the other tests

**Review by sultan-duisenbay — CHANGES_REQUESTED** [review]
left reviews

File: pkg/updater/resultstore/resultstore.go [hunk 17]
```diff
@@ -453,6 +456,85 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
...
+		if testMethodLimit == 0 {
+			testMethodLimit = defaultTestMethodLimit
+		}
+	}
+	return testMethodLimit
+}
+
+func testResultStatus(testResult *resultstorepb.Test) statuspb.TestStatus {
+	return statuspb.TestStatus(testResult.GetTestCase().Result)
+}
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:502]: you have to perform explicit mapping between TestCase results and Testgrid test statuses. e.g.

skipped = pass with skips, suppressed = pass with skips, etc. refer to :


```
func mapStatus(status testsuitepb.TestCase_Result) TestResultStatus {
	switch status {
	case testsuitepb.TestCase_COMPLETED:
		return TestCompleted
	case testsuitepb.TestCase_INTERRUPTED:
		return TestInterrupted
	case testsuitepb.TestCase_CANCELLED:
		return TestCancelled
	case testsuitepb.TestCase_FILTERED:
		return TestFiltered
	case testsuitepb.TestCase_SKIPPED:
		return TestSkipped
	case testsuitepb.TestCase_SUPPRESSED:
		return TestSuppressed
	default:
		return TestResultUnspecified
	}
}
```

and
```
case res == TestSkipped:
							trCell.Result = tspb.TestStatus_PASS_WITH_SKIPS
						case res == TestSuppressed:
							trCell.Result = tspb.TestStatus_PASS_WITH_SKIPS
						case res == TestCancelled:
							trCell.Result = tspb.TestStatus_CANCEL
						case res == TestInterrupted:
							trCell.Result = tspb.TestStatus_CANCEL
						case testResult.Failures() > 0 || testResult.Errors() > 0:
							trCell.Result = tspb.TestStatus_FAIL
						default:
							trCell.Result = tspb.TestStatus_PASS
```

**Review by sultan-duisenbay — CHANGES_REQUESTED** [review]

File: pkg/updater/resultstore/resultstore.go (hunk 17 above)
Comment (sultan-duisenbay) [reply pkg/updater/resultstore/resultstore.go:502]: also add a couple of test cases to processGroup, that function is affected by your changes now

Comment (bryan-lou) [reply pkg/updater/resultstore/resultstore.go]: done

File: pkg/updater/resultstore/resultstore.go [hunk 18]
```diff
@@ -39,6 +40,25 @@ import (
...
+	// TestCancelled is v2.TestCase.CANCELLED.
+	TestCancelled = 3
+	// TestFiltered is v2.TestCase.FILTERED.
+	TestFiltered = 4
+	// TestSkipped is v2.TestCase.SKIPPED.
+	TestSkipped = 5
+	// TestSuppressed is v2.TestCase.SUPPRESSED.
+	TestSuppressed = 6
+)
+
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:61]: again - we don't need this intermediary translation layer, please convert it directly in your `mapStatusToCellResult` function below

File: pkg/updater/resultstore/resultstore.go [hunk 19]
```diff
@@ -453,6 +475,101 @@ func processGroup(tg *configpb.TestGroup, group *invocationGroup) *updater.Infla
...
+		return statuspb.TestStatus_PASS_WITH_SKIPS
+	case res == TestCancelled:
+		return statuspb.TestStatus_CANCEL
+	case res == TestInterrupted:
+		return statuspb.TestStatus_CANCEL
+	case len(testCase.Failures) > 0 || len(testCase.Errors) > 0:
+		return statuspb.TestStatus_FAIL
+	default:
+		return statuspb.TestStatus_PASS
+	}
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:536]: perform the mapping directly from resultstore status to cell status

**Review by sultan-duisenbay — CHANGES_REQUESTED** [review]

File: pkg/updater/resultstore/resultstore.go [hunk 20]
```diff
@@ -39,6 +40,8 @@ import (
 // check if interface is implemented correctly
 var _ updater.TargetResult = &singleActionResult{}
 
+type TestResultStatus int64
+
```
Comment (sultan-duisenbay) [inline pkg/updater/resultstore/resultstore.go:44]: we don't need this now

**Review by sultan-duisenbay — CHANGES_REQUESTED** [review]
also squash the commits
/hold

**Pull request merged**